test/performance/range/gap_decorator_benchmark.cpp

```diff
@@ -0,0 +1,452 @@
+ * seq_len      final sequence length
+ */
+template<typename size_type, typename sequence_type>
+void resize(std::vector<size_type> & gaps, sequence_type & seq, unsigned int seq_len)
+{
+    size_type letter_acc = 0;
+    size_type gap_pos = 0;
+    size_type gap_acc = 0;
+
+    while (gap_pos < gaps.size() && gap_acc + letter_acc < seq_len)
+    {
+        if (!gaps[gap_pos])
+        {
+            ++letter_acc;
+        }
+        else
+        {
+            if (letter_acc + gap_acc + gaps[gap_pos] > seq_len)
+            {
+                gaps[gap_pos] = seq_len - gap_acc - letter_acc;
+                gap_acc += gaps[gap_pos];
+                ++gap_pos;
+                break;
+            }
+            else
+            {
+                gap_acc += gaps[gap_pos];
+            }
+        }
+        ++gap_pos;
+    }
+    seq.resize(std::max<size_type>(1, letter_acc));  // resize ungapped sequence
+    gaps.resize(gap_pos);      // trim sampled gap vector
+}
+
+/* Helper function to prepare a gapped sequence for the benchmark (case gap_flag=true)
+ *
+ * Parameters:
+ * size_type        size type of the sequence
+ * gap_decorator_t  gap decorator type, e.g. gap_decorator_anchor_set
+ *
+ * Arguments:
+ * gaps             reference to gap vector
+ * gap_decorator    reference to gap decorator
+ */
+template<typename gap_decorator_t>
+void insert_gaps(std::vector<typename gap_decorator_t::size_type> & gaps, gap_decorator_t & gap_decorator)
+{
+    typename gap_decorator_t::size_type gap_acc = 0;
+    for (typename gap_decorator_t::size_type i = 0; i < gaps.size(); ++i)
+    {
+        if (gaps[i])
+        {
+            auto it = std::ranges::begin(gap_decorator);
+            std::ranges::advance(it, std::min(i + gap_acc, gap_decorator.size()));
+            insert_gap(gap_decorator, it, gaps[i]);
+        }
+        gap_acc += gaps[i];
+    }
+}
+
+// ============================================================================
+//  read left to right (looped in case #ops exceeds sequence length)
+// ============================================================================
+/* Parameters:
+ * gap_decorator_t      gap decorator class, e.g. gap_decorator_anchor_set
+ * gapped_flag          operate on already gapped (true) or ungapped sequence (false)
+ */
+template <typename gap_decorator_t, bool gapped_flag>
+static void read_left2right(benchmark::State& state)
+{
+    // get target sequence length from current range state
+    unsigned int seq_len = state.range(0);
+    using size_type = typename gap_decorator_t::size_type;
+    using sequence_type = remove_cvref_t<detail::unaligned_seq_t<gap_decorator_t>>;
+    sequence_type seq(seq_len, 'A'_dna4);
+
+    // vector of sampled gap lengths for each position
+    std::vector<size_type> gaps(seq_len, 0);
+
+    // determine sum of gaps and non-gap symbols for not exceeding targeted sequence length
+    if constexpr(gapped_flag)
+    {
+        sample<size_type>(gaps, seq_len);
```

Comment:
```suggestion
        sample(gaps, seq_len);
```
I think the size_type is auto-deducted so you do not need it here and also not in the resize call :)

Reply:
Hmm, didn't work out. Looks like I have to explicitly write it, the compiler is not able to deduce it.